warehouse/email/__init__.py

```diff
@@ -56,8 +74,18 @@ def _send_email_to_user(request, user, msg, *, email=None, allow_unverified=Fals
     if email is None or not (email.verified or allow_unverified):
         return
 
+    # We should only store/display IP address of an 'email sent' event if the user
+    # who triggered the email event is the one who receives the email. Else display
+    # 'Redacted' to prevent user privacy concerns
+    user_email = request.db.query(Email).filter(Email.email == email.email).one()
+    redact_ip = user_email.user_id != request.user.id
```

Comment:
We need to handle the case where `request.user` is `None` (for example, when requesting a password reset)

Reply:
Hmm, I printed `request.user` in the console and it seems to have a valid value during password reset and it also displays the appropriate event, are there any other scenarios that `request.user` can be `None`? If so, should we marked it as redacted since we can't verify the source?

![image](https://user-images.githubusercontent.com/27223486/86860565-b0a75e80-c08a-11ea-95a7-374486d8cb4a.png)

Reply:
Here's the full stacktrace I get with this branch when requesting a password reset:

```
Traceback (most recent call last):
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid_debugtoolbar/panels/performance.py", line 58, in resource_timer_handler
    result = handler(request)
  File "/opt/warehouse/src/warehouse/raven.py", line 40, in raven_tween
    return handler(request)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid_tm/__init__.py", line 178, in tm_tween
    reraise(*exc_info)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid_tm/compat.py", line 36, in reraise
    raise value
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid_tm/__init__.py", line 143, in tm_tween
    response = handler(request)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/tweens.py", line 43, in excview_tween
    response = _error_handler(request, exc)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/tweens.py", line 17, in _error_handler
    reraise(*exc_info)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/compat.py", line 179, in reraise
    raise value
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File "/opt/warehouse/src/warehouse/cache/http.py", line 74, in conditional_http_tween
    response = handler(request)
  File "/opt/warehouse/src/warehouse/sanity.py", line 76, in sanity_tween_egress
    return unicode_redirects(handler(request))
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/router.py", line 147, in handle_request
    response = _call_view(
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/view.py", line 667, in _call_view
    response = view_callable(context, request)
  File "/opt/warehouse/src/warehouse/cache/http.py", line 33, in wrapped
    return view(context, request)
  File "/opt/warehouse/src/warehouse/cache/http.py", line 33, in wrapped
    return view(context, request)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/viewderivers.py", line 514, in csrf_view
    return view(context, request)
  File "/opt/warehouse/src/warehouse/metrics/views.py", line 34, in wrapper_view
    return view(context, request)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
    result = view(context, request)
  File "/opt/warehouse/lib/python3.8/site-packages/pyramid/viewderivers.py", line 144, in _requestonly_view
    response = view(request)
  File "/opt/warehouse/src/warehouse/accounts/views.py", line 541, in request_password_reset
    send_password_reset_email(request, (user, email))
  File "/opt/warehouse/src/warehouse/email/__init__.py", line 141, in wrapper
    _send_email_to_user(
  File "/opt/warehouse/src/warehouse/email/__init__.py", line 81, in _send_email_to_user
    redact_ip = user_email.user_id != request.user.id
AttributeError: 'NoneType' object has no attribute 'id'
```

The `request.user` attribute is `None` whenever there isn't a logged-in user.

Reply:
Ah I see, `request.user` is `None` when you reset password during login. I assumed when resetting under `Account settings`. Good catch! I will fix that.